src/inmanta/deploy/work.py

```diff
@@ -345,7 +358,23 @@ def deploy_with_context(
         for resourceId in queued:
             if resourceId in resources:
                 task = tasks.Deploy(resource=resourceId)
-                self.agent_queues._in_progress[task] = min(self.agent_queues._in_progress[task], priority)
+                assert task in self.agent_queues._in_progress
+                actual_task = self.agent_queues.retrieve_task_in_progress(task)
+                assert actual_task, "The task retrieved from the queue should exist"
+                if self.agent_queues._in_progress[task] > priority:
+                    self.agent_queues._in_progress[task] = priority
+                    actual_task.set_reason(reason=reason)
+                elif self.agent_queues._in_progress[task] < priority or (
+                    self.agent_queues._in_progress[task] == priority and actual_task.reason == reason
+                ):
+                    # We don't need to do anything in that particular case, the lowest priority is already set
+                    pass
+                else:
+                    raise RuntimeError(
+                        f"Two different priorities with the same weight have been associated with the current task: "
+                        f"Old priority -> (`{self.agent_queues._in_progress[task]}`, reason: {actual_task.reason}) "
+                        f"- New priority -> (`{priority}`, reason: {reason})"
+                    )
```

Comment:
We very explicitly don't care of two different priorities with the same weight have been associated with the current task and definitely don't want to raise an exception then, as it is expected to happen